rclrs/src/time_source.rs

```diff
@@ -0,0 +1,180 @@
+/// This struct instance can be created via [`TimeSource::builder()`][2].
+///
+/// The default values for optional fields are:
+/// - `clock_qos: QOS_PROFILE_CLOCK`[3]
+/// - `use_clock_thread: true`
+///
+///
+/// [1]: crate::TimeSource
+/// [2]: crate::TimeSource::builder
+/// [3]: crate::QOS_PROFILE_CLOCK
+pub struct TimeSourceBuilder {
+    node: Weak<Node>,
+    clock_qos: QoSProfile,
+    use_clock_thread: bool,
+    clock_type: ClockType,
+}
+
+impl TimeSourceBuilder {
+    /// Creates a builder for a time source that drives the given clock for the given node.
+    pub fn new(node: Weak<Node>, clock_type: ClockType) -> Self {
+        Self {
+            node,
+            clock_qos: QOS_PROFILE_CLOCK,
+            use_clock_thread: true,
+            clock_type,
+        }
+    }
+
+    /// Sets the QoS for the `/clock` topic.
+    pub fn clock_qos(mut self, clock_qos: QoSProfile) -> Self {
+        self.clock_qos = clock_qos;
+        self
+    }
+
+    /// Sets use_clock_thread.
+    ///
+    /// If set to `true`, the clock callbacks will run in a separate thread.
+    /// NOTE: Currently unimplemented
+    pub fn use_clock_thread(mut self, use_clock_thread: bool) -> Self {
+        self.use_clock_thread = use_clock_thread;
+        self
+    }
+
+    /// Builds the `TimeSource` and attaches the provided `Node` and `Clock`.
+    pub fn build(self) -> Result<TimeSource, RclrsError> {
+        let clock = match self.clock_type {
+            ClockType::RosTime | ClockType::SystemTime => Clock::system(),
+            ClockType::SteadyTime => Clock::steady(),
+        }?;
+        let mut source = TimeSource {
+            _node: Weak::new(),
+            _clock: RwLock::new(clock),
+            _clock_source: Arc::new(Mutex::new(None)),
+            _requested_clock_type: self.clock_type,
+            _clock_qos: self.clock_qos,
+            _use_clock_thread: self.use_clock_thread,
+            _clock_subscription: None,
+            _last_received_time: Arc::new(Mutex::new(None)),
+        };
+        source.attach_node(self.node);
+        Ok(source)
+    }
+}
+
+impl TimeSource {
+    /// Creates a new `TimeSource` with default parameters.
+    pub fn new(node: Weak<Node>, clock_type: ClockType) -> Self {
+        TimeSourceBuilder::new(node, clock_type).build().unwrap()
+    }
+
+    /// Creates a new `TimeSourceBuilder` with default parameters.
+    pub fn builder(node: Weak<Node>, clock_type: ClockType) -> TimeSourceBuilder {
+        TimeSourceBuilder::new(node, clock_type)
+    }
+
+    /// Returns the clock that this TimeSource is controlling.
+    pub fn get_clock(&self) -> Clock {
+        self._clock.read().unwrap().clone()
+    }
+
+    /// Attaches the given node to to the `TimeSource`, using its interface to read the
+    /// `use_sim_time` parameter and create the clock subscription.
+    pub fn attach_node(&mut self, node: Weak<Node>) {
+        self._node = node;
```

Comment:
Perhaps I missed something, but do we actually need to store the weak reference to the Node at all here? It seems like we only read the `use_sim_time` parameter at the time of a `TimeSource` structs creation. Unless we plan to re-use these `TimeSource`'s or if it actually does make sense to have them update when the parameter updates, I don't think we should store the weak reference.

Reply:
Indeed we only read the value of the parameter (and create the clock subscriber) once at creation time right now so we could technically just pass the node as an argument to the builder and just store an `Option<Subscription>` instead.

However, this behavior is not the full implementation and all other client libraries do store the node in the time source. The reason for this is that the time source needs to be able to subscribe to parameter changes and, when the `use_sim_time` parameter changes at runtime, be able to create and destroy subscriptions. This is not possible without holding a reference to the node.
If we just did the "bare minimum" now and didn't pass / store the `node` we would have to change the API once parameters are implemented.

Reply:
I wonder how much access an end user actually needs to `TimeSource`. I can't remember ever needing to touch that as a user in any other client library.

Is it feasible to slap `pub(crate)` on enough structs and methods that we can avoid API commitments until the full implementation is ready?

Reply:
Thanks for the suggestion! Indeed you are right and that part of  the code is very rarely touched by users.
In https://github.com/ros2-rust/ros2_rust/pull/325/commits/382f780edf87a5190d1fdc3b3ea14d9bd3a25a22 I made everything in `time_source.rs` either private or `pub(crate)` so it doesn't export anything as a public API. I also removed the `use_clock_thread` builder argument since it is not implemented yet and sounds like might not be for some time.
This added another unwrap that however falls in the class of the ones I described above (only fail if we can't allocate a few bytes of memory) so it's mostly noise.

Reply:
> when the `use_sim_time` parameter changes at runtime

This doesn't happen currently, I'm guessing its blocked by the lack of a parameter implementation in this PR?

I would prefer if we did something similar to the parameters impl where we deviate slightly from rclcpp. Forcing end users to wrap all nodes in `Arc` is a large API change for a niche runtime configuration option IMO

But I'll defer to @esteve, @nnmm, and @jhdcs. I think this PR has been open long enough and we can always iterate :)

Reply:
Realistically I think it'll be close to impossible to implement a whole bunch of sophisticated features down the road without requiring that `Node`s are wrapped in `Arc` unless we take an approach where almost every field inside in the `Node` struct is wrapped in its own `Arc` to hide this implementation detail.

But I don't think that's a great idea because it just hides important information from the user about the lifecycle and identity of a node. For example a user might mistakenly think that they can create a second node instance by cloning an existing one when in reality they would only be getting a new reference to the same node instance.

Putting the Arc on the outside explicitly communicates the lifecycle and shareability properties of the node.

I can imagine implementing a Rust client library that does not use Arc for nodes if it were implemented inside of an ECS framework, but I think such a library should be done as a separate project from this one.